src/main/scala/loamstream/loam/intake/aggregator/SourceColumns.scala

```diff
@@ -5,32 +5,57 @@ import loamstream.loam.intake.ColumnName
 final case class SourceColumns(
     marker: ColumnName,
     pValue: ColumnName,
-    zScore: ColumnName,
-    stderr: ColumnName,
-    beta: ColumnName,
-    oddsRatio: ColumnName,
-    eaf: ColumnName,
-    maf: ColumnName) {
+    zScore: Option[ColumnName] = None,
+    stderr: Option[ColumnName] = None,
+    beta: Option[ColumnName] = None,
+    oddsRatio: Option[ColumnName] = None,
+    eaf: Option[ColumnName] = None,
+    maf: Option[ColumnName] = None) {
   
-  private val mapping: Map[ColumnName, ColumnName] = Map(
+  validate()
+  
+  private val mapping: Map[ColumnName, ColumnName] = {
+    //mandatory columns
+    val mandatory = Map(
       ColumnNames.marker -> this.marker,
-      ColumnNames.pvalue -> this.pValue,
-      ColumnNames.zscore -> this.zScore,
-      ColumnNames.stderr -> this.stderr,
-      ColumnNames.beta -> this.beta,
-      ColumnNames.odds_ratio -> this.oddsRatio,
-      ColumnNames.eaf -> this.eaf,
-      ColumnNames.maf -> this.maf)
+      ColumnNames.pvalue -> this.pValue)
       
-  def asConfigFileContents: String = {
-    val aggregatorColumnNameToSourceColumnNameLines = for {
-      aggregatorColumnName <- ColumnNames.values
-    } yield {
-      val sourceColumnName = mapping(aggregatorColumnName)
+    mandatory ++
+      stderr.map(ColumnNames.stderr -> _) ++
+      zScore.map(ColumnNames.zscore -> _) ++
+      beta.map(ColumnNames.beta -> _) ++
+      oddsRatio.map(ColumnNames.odds_ratio -> _) ++
+      eaf.map(ColumnNames.eaf -> _) ++
+      maf.map(ColumnNames.maf -> _)
+  }
       
-      s"${aggregatorColumnName.name} ${sourceColumnName.name}"
+  def asConfigFileContents: String = {
+    val aggregatorColumnNameToSourceColumnNameLines = mapping.map {
+      case (aggregatorColumnName, sourceColumnName) => s"${aggregatorColumnName.name} ${sourceColumnName.name}"
     }
     
     aggregatorColumnNameToSourceColumnNameLines.mkString(System.lineSeparator)
   }
+  
+  private def validate(): Unit = {
```

Comment:
There's always the question of explicit vs. implicit. I'm just wondering how this code would compare in the future (if it grows) vs. a mapping that's iterated and tested?

```scala
// the validation map could exist anywhere in the code maybe?
val validationMap: Map[Option[Double], Seq[Option[Double]]] = Map(
  maf -> Seq(eaf),
  beta -> Seq(oddsRatio),
  oddsRation -> Seq(beta),
  stderr -> Seq(beta, pValue),
  zscore -> Seq(beta, stderr),
)

for ((col, deps) <- validationMap) {
  if (col.isEmpty) {
    require(deps.forall(_.isDefined))
  }
}
```

Reply:
Just noting for myself that one possible drawback to the above code is maybe if `require` throws an error, it will be unclear which column is actually missing?

Reply:
I'll think this over.  Duplicating the logic from the aggregator side is doomed to fail eventually, but I wanted some way to fail fast if what users wrote was definitely not going to work.

Reply:
If any validation logic like this stays in the future, I'll probably do something like this in a future PR.